File: examples/Lua/tasksink2.lua
--
--  Task sink - design 2
--  Adds pub-sub flow to send kill signal to workers
--
--  Author: Robert G. Jakabosky <bobby@sharedrealm.com>
--
require"zmq"
require"zhelpers"
local fmod = math.fmod

local context = zmq.init(1)

--  Socket to receive messages on
local receiver = context:socket(zmq.PULL)
receiver:bind("tcp://*:5558")

--  Socket for worker control
local controller = context:socket(zmq.PUB)
controller:bind("tcp://*:5559")

--  Wait for start of batch
local msg = receiver:recv()

--  Start our clock now
local start_time = s_clock ()

--  Process 100 confirmations
local task_nbr
for task_nbr=0,99 do
    local msg = receiver:recv()

    if (fmod(task_nbr, 10) == 0) then
        printf (":")
    else
        printf (".")
    end
    io.stdout:flush()
end
printf("Total elapsed time: %d msec\n", (s_clock () - start_time))

--  Send kill signal to workers
controller:send("KILL")

--  Finished
s_sleep (1000)              --  Give 0MQ time to deliver

receiver:close()
controller:close()
context:term()

File: examples/OCaml/tasksink2.ml
(**
 * Task sink - design 2
 * Adds pub-sub flow to send kill signal to workers
 *)
open Zmq
open Helpers

let () =
  with_context @@ fun ctx ->
  (* Socket to receive messages on *)
  with_socket ctx Socket.pull @@ fun receiver ->
  Socket.bind receiver "tcp://*:5558";

  (* Scoket for worker control *)
  with_socket ctx Socket.pub @@ fun controller ->
  Socket.bind controller "tcp://*:5559";

  (* Wait for start of batch *)
  let _ = Socket.recv receiver in

  (* Start our clock now *)
  let start_time = clock_ms () in

  (* Process 100 confirmations *)
  for taskNum = 0 to 99 do
    let _ = Socket.recv receiver in
    printfn @@ if ((taskNum / 10) * 10 == taskNum) then ":" else ".";
  done;

  printfn "Total elapsed time: %d msec" (clock_ms () - start_time);

  (* Send kill signal to workers *)
  Socket.send controller "KILL";
